File: src/image-featurizer/src/main/scala/ImageLIME.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark

import breeze.linalg.{DenseMatrix => BDM}
import com.microsoft.ml.spark.FluentAPI._
import com.microsoft.ml.spark.schema.DatasetExtensions
import org.apache.spark.internal.{Logging => SLogging}
import org.apache.spark.ml.linalg.SQLDataTypes.{MatrixType, VectorType}
import org.apache.spark.ml.linalg.{DenseMatrix, DenseVector}
import org.apache.spark.ml.param._
import org.apache.spark.ml.util.{ComplexParamsReadable, ComplexParamsWritable, Identifiable}
import org.apache.spark.ml.{LimeNamespaceInjections, Transformer}
import org.apache.spark.sql.catalyst.encoders.RowEncoder
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.sql.{DataFrame, Dataset, Row}

import scala.collection.mutable

object LIMEUtils extends SLogging {

  def localAggregateBy(df: DataFrame, groupByCol: String, colsToSquish: Seq[String]): DataFrame = {
    val schema = new StructType(df.schema.fields.map {
      case field if colsToSquish.contains(field.name) => StructField(field.name, ArrayType(field.dataType))
      case f => f
    })
    val encoder = RowEncoder(schema)
    val indiciesToSquish = colsToSquish.map(df.schema.fieldIndex)
    df.mapPartitions { it =>
      val isEmpty = it.isEmpty
      if (isEmpty) {
        (Nil: Seq[Row]).toIterator
      } else {
        // Current Id, What we have accumulated, Previous Row
        val accumulator = mutable.ListBuffer[Seq[Any]]()
        var currentId: Option[Any] = None
        var prevRow: Option[Row] = None

        def returnState(accumulated: List[Seq[Any]], prevRow: Row): Row = {
          Row.fromSeq(prevRow.toSeq.zipWithIndex.map {
            case (v, i) if indiciesToSquish.contains(i) =>
              accumulated.map(_.apply(indiciesToSquish.indexOf(i)))
            case (v, i) => v
          })
        }

        def enqueueAndMaybeReturn(row: Row): Option[Row] = {
          val id = row.getAs[Any](groupByCol)
          if (currentId.isEmpty) {
            currentId = Some(id)
            prevRow = Some(row)
            accumulator += colsToSquish.map(row.getAs[Any])
            None
          } else if (id != currentId.get) {
            val accumulated = accumulator.toList
            accumulator.clear()
            accumulator += colsToSquish.map(row.getAs[Any])
            val modified = returnState(accumulated, prevRow.get)
            currentId = Some(id)
            prevRow = Some(row)
            Some(modified)
          } else {
            prevRow = Some(row)
            accumulator += colsToSquish.map(row.getAs[Any])
            None
          }
        }

        val it1 = it
          .flatMap(row => enqueueAndMaybeReturn(row))
          .map(r => Left(r): Either[Row, Null])
          .++(Seq(Right(null): Either[Row, Null]))

        val ret = it1.map {
          case Left(r) => r: Row
          case Right(_) => returnState(accumulator.toList, prevRow.getOrElse {
            logWarning("Could not get previous row in local aggregator, this is an error that should be fixed")
            null
          }): Row
        }
        ret
      }

    }(encoder)
  }
}

object ImageLIME extends ComplexParamsReadable[ImageLIME]

/** Distributed implementation of
  * Local Interpretable Model-Agnostic Explanations (LIME)
  *
  * https://arxiv.org/pdf/1602.04938v1.pdf
  */
class ImageLIME(val uid: String) extends Transformer
  with HasInputCol with HasOutputCol with HasLabelCol
  with Wrappable with ComplexParamsWritable
  with HasModifier with HasCellSize {

  def this() = this(Identifiable.randomUID("LIME"))

  val model = new TransformerParam(this, "model", "Model to try to locally approximate")

  def getModel: Transformer = $(model)

  def setModel(v: Transformer): this.type = set(model, v)

  val nSamples = new IntParam(this, "nSamples", "The number of samples to generate")

  def getNSamples: Int = $(nSamples)

  def setNSamples(v: Int): this.type = set(nSamples, v)

  val samplingFraction = new DoubleParam(this, "samplingFraction", "The fraction of superpixels to keep on")

  def getSamplingFraction: Double = $(samplingFraction)

  def setSamplingFraction(d: Double): this.type = set(samplingFraction, d)

  val superpixelCol = new Param[String](this, "superpixelCol", "The column holding the superpixel decompositions")

  def getSuperpixelCol: String = $(superpixelCol)

  def setSuperpixelCol(v: String): this.type = set(superpixelCol, v)

  val regularization = new DoubleParam(this, "regularization", "regularization param for the lasso")

  def getRegularization: Double = $(regularization)

  def setRegularization(v: Double): this.type = set(regularization, v)

  setDefault(nSamples -> 900, cellSize -> 16, modifier -> 130, regularization -> 0.0,
    samplingFraction -> 0.3, superpixelCol -> "superpixels")

  override def transform(dataset: Dataset[_]): DataFrame = {

    val df = dataset.toDF
    val model = getModel

    val idCol = DatasetExtensions.findUnusedColumnName("imageID", df)
    val statesCol = DatasetExtensions.findUnusedColumnName("states", df)
    val inputCol2 = DatasetExtensions.findUnusedColumnName("inputCol2", df)

    // Data frame with new column containing superpixels (Array[Cluster]) for each row (image)
    val spt = new SuperpixelTransformer()
      .setCellSize(getCellSize)
      .setModifier(getModifier)
      .setInputCol(getInputCol)
      .setOutputCol(getSuperpixelCol)

    val spDF = spt.transform(df)

    // Indices of the columns containing each image and image's superpixels
    val inputType = df.schema(getInputCol).dataType
    val censorUDF = inputType match {
      case BinaryType => Superpixel.censorBinaryUDF
      case t if ImageSchemaUtils.isImage(t) => Superpixel.censorUDF
    }

    def getSamples(n: Int): Seq[Seq[Boolean]] = {
      Superpixel
        .clusterStateSampler(getSamplingFraction, n)
        .take(getNSamples).map(_.toSeq).toSeq
    }

    val getSampleUDF = udf(getSamples _, ArrayType(ArrayType(BooleanType)))

    val fitLassoUDF = udf(LimeNamespaceInjections.fitLasso _, VectorType)

    def arrToMat(dvs: Seq[DenseVector]): DenseMatrix = {
      val mat = BDM(dvs.map(_.values): _*)
      new DenseMatrix(mat.rows, mat.cols, mat.data)
    }

    val arrToMatUDF = udf(arrToMat _, MatrixType)

    def arrToVect(ds: Seq[Double]): DenseVector = {
      new DenseVector(ds.toArray)
    }

    val arrToVectUDF = udf(arrToVect _, VectorType)

    val mapped = spDF.withColumn(idCol, monotonically_increasing_id())
      .withColumnRenamed(getInputCol, inputCol2)
      .withColumn(statesCol, explode_outer(getSampleUDF(size(col(getSuperpixelCol).getField("clusters")))))
      .withColumn(getInputCol, censorUDF(col(inputCol2), col(spt.getOutputCol), col(statesCol)))
      .withColumn(statesCol, udf(
        { barr: Seq[Boolean] => new DenseVector(barr.map(b => if (b) 1.0 else 0.0).toArray) },
        VectorType)(col(statesCol)))
      .mlTransform(model)
      .drop(getInputCol)

    LIMEUtils.localAggregateBy(mapped, idCol, Seq(statesCol, getLabelCol))
      .withColumn(statesCol, arrToMatUDF(col(statesCol)))
      .withColumn(getLabelCol, arrToVectUDF(col(getLabelCol)))
      .withColumn(getOutputCol, fitLassoUDF(col(statesCol), col(getLabelCol), lit(getRegularization)))
      .drop(statesCol, getLabelCol)
      .withColumnRenamed(inputCol2, getInputCol)
  }

  override def copy(extra: ParamMap): Transformer = defaultCopy(extra)

  override def transformSchema(schema: StructType): StructType = {
    schema.add(getSuperpixelCol, SuperpixelData.schema).add(getOutputCol, VectorType)
  }

}

File: src/src/main/resources/mmlspark/ImageLIME.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *
from mmlspark.TypeConversionUtils import generateTypeConverter, complexTypeConverter

@inherit_doc
class ImageLIME(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """


    Args:

        cellSize (double): Number that controls the size of the superpixels (default: 16.0)
        inputCol (str): The name of the input column
        model (object): Model to try to locally approximate
        modifier (double): Controls the trade-off spatial and color distance (default: 130.0)
        nSamples (int): The number of samples to generate (default: 900)
        outputCol (str): The name of the output column
        predictionCol (str): prediction column name (default: prediction)
        regularization (double): regularization param for the lasso (default: 0.0)
        samplingFraction (double): The fraction of superpixels to keep on (default: 0.3)
        superpixelCol (str): The column holding the superpixel decompositions (default: superpixels)
    """

    @keyword_only
    def __init__(self, cellSize=16.0, inputCol=None, model=None, modifier=130.0, nSamples=900, outputCol=None, predictionCol="prediction", regularization=0.0, samplingFraction=0.3, superpixelCol="superpixels"):
        super(ImageLIME, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.ImageLIME")
        self._cache = {}
        self.cellSize = Param(self, "cellSize", "cellSize: Number that controls the size of the superpixels (default: 16.0)")
        self._setDefault(cellSize=16.0)
        self.inputCol = Param(self, "inputCol", "inputCol: The name of the input column")
        self.model = Param(self, "model", "model: Model to try to locally approximate", generateTypeConverter("model", self._cache, complexTypeConverter))
        self.modifier = Param(self, "modifier", "modifier: Controls the trade-off spatial and color distance (default: 130.0)")
        self._setDefault(modifier=130.0)
        self.nSamples = Param(self, "nSamples", "nSamples: The number of samples to generate (default: 900)")
        self._setDefault(nSamples=900)
        self.outputCol = Param(self, "outputCol", "outputCol: The name of the output column")
        self.predictionCol = Param(self, "predictionCol", "predictionCol: prediction column name (default: prediction)")
        self._setDefault(predictionCol="prediction")
        self.regularization = Param(self, "regularization", "regularization: regularization param for the lasso (default: 0.0)")
        self._setDefault(regularization=0.0)
        self.samplingFraction = Param(self, "samplingFraction", "samplingFraction: The fraction of superpixels to keep on (default: 0.3)")
        self._setDefault(samplingFraction=0.3)
        self.superpixelCol = Param(self, "superpixelCol", "superpixelCol: The column holding the superpixel decompositions (default: superpixels)")
        self._setDefault(superpixelCol="superpixels")
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, cellSize=16.0, inputCol=None, model=None, modifier=130.0, nSamples=900, outputCol=None, predictionCol="prediction", regularization=0.0, samplingFraction=0.3, superpixelCol="superpixels"):
        """
        Set the (keyword only) parameters

        Args:

            cellSize (double): Number that controls the size of the superpixels (default: 16.0)
            inputCol (str): The name of the input column
            model (object): Model to try to locally approximate
            modifier (double): Controls the trade-off spatial and color distance (default: 130.0)
            nSamples (int): The number of samples to generate (default: 900)
            outputCol (str): The name of the output column
            predictionCol (str): prediction column name (default: prediction)
            regularization (double): regularization param for the lasso (default: 0.0)
            samplingFraction (double): The fraction of superpixels to keep on (default: 0.3)
            superpixelCol (str): The column holding the superpixel decompositions (default: superpixels)
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setCellSize(self, value):
        """

        Args:

            cellSize (double): Number that controls the size of the superpixels (default: 16.0)

        """
        self._set(cellSize=value)
        return self


    def getCellSize(self):
        """

        Returns:

            double: Number that controls the size of the superpixels (default: 16.0)
        """
        return self.getOrDefault(self.cellSize)


    def setInputCol(self, value):
        """

        Args:

            inputCol (str): The name of the input column

        """
        self._set(inputCol=value)
        return self


    def getInputCol(self):
        """

        Returns:

            str: The name of the input column
        """
        return self.getOrDefault(self.inputCol)


    def setModel(self, value):
        """

        Args:

            model (object): Model to try to locally approximate

        """
        self._set(model=value)
        return self


    def getModel(self):
        """

        Returns:

            object: Model to try to locally approximate
        """
        return self._cache.get("model", None)


    def setModifier(self, value):
        """

        Args:

            modifier (double): Controls the trade-off spatial and color distance (default: 130.0)

        """
        self._set(modifier=value)
        return self


    def getModifier(self):
        """

        Returns:

            double: Controls the trade-off spatial and color distance (default: 130.0)
        """
        return self.getOrDefault(self.modifier)


    def setNSamples(self, value):
        """

        Args:

            nSamples (int): The number of samples to generate (default: 900)

        """
        self._set(nSamples=value)
        return self


    def getNSamples(self):
        """

        Returns:

            int: The number of samples to generate (default: 900)
        """
        return self.getOrDefault(self.nSamples)


    def setOutputCol(self, value):
        """

        Args:

            outputCol (str): The name of the output column

        """
        self._set(outputCol=value)
        return self


    def getOutputCol(self):
        """

        Returns:

            str: The name of the output column
        """
        return self.getOrDefault(self.outputCol)


    def setPredictionCol(self, value):
        """

        Args:

            predictionCol (str): prediction column name (default: prediction)

        """
        self._set(predictionCol=value)
        return self


    def getPredictionCol(self):
        """

        Returns:

            str: prediction column name (default: prediction)
        """
        return self.getOrDefault(self.predictionCol)


    def setRegularization(self, value):
        """

        Args:

            regularization (double): regularization param for the lasso (default: 0.0)

        """
        self._set(regularization=value)
        return self


    def getRegularization(self):
        """

        Returns:

            double: regularization param for the lasso (default: 0.0)
        """
        return self.getOrDefault(self.regularization)


    def setSamplingFraction(self, value):
        """

        Args:

            samplingFraction (double): The fraction of superpixels to keep on (default: 0.3)

        """
        self._set(samplingFraction=value)
        return self


    def getSamplingFraction(self):
        """

        Returns:

            double: The fraction of superpixels to keep on (default: 0.3)
        """
        return self.getOrDefault(self.samplingFraction)


    def setSuperpixelCol(self, value):
        """

        Args:

            superpixelCol (str): The column holding the superpixel decompositions (default: superpixels)

        """
        self._set(superpixelCol=value)
        return self


    def getSuperpixelCol(self):
        """

        Returns:

            str: The column holding the superpixel decompositions (default: superpixels)
        """
        return self.getOrDefault(self.superpixelCol)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.ImageLIME"

    @staticmethod
    def _from_java(java_stage):
        module_name=ImageLIME.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".ImageLIME"
        return from_java(java_stage, module_name)
